### PR TITLE
SQL-1123: Implement diagnostics and error handling

### PR DESCRIPTION
**Pull request by bucaojit**
Added diagnostics module from connector samples. 
Added Logging of configuration information.
Implemented OnError function for additional DataSource error handling.

Example error message:
```
Unable to connect
 
We encountered an error while trying to connect.
Details: "Data source error occurred.
SQLSTATE: 01000
NativeError: 444
Error message: ODBC: ERROR [01000] The driver returned invalid (or failed to return) SQL_DRIVER_ODBC_VER: 03.80"
```
Example Diagnostic message:
```
DataMashup.Trace Information: 24580 : {"Start":"2023-02-15T18:31:34.7763825Z","Action":"Engine/Module/MongoDBAtlasODBC/IO/MongoDBAtlasODBC","HostProcessId":"36944","PartitionKey":"Section1/Query1/Source","Message":"BuildOdbcConfig: [ AstVisitor = [ LimitClause = (skip as any, take as any) as any => (...)  ] , HierarchicalNavigation = true, SqlCapabilities = [ FractionalSecondsScale = 3, LimitClauseKind = 2, SupportsDerivedTable = true, SupportsNumericLiterals = true, SupportsStringLiterals = true, SupportsOdbcDateLiterals = true, SupportsOdbcTimeLiterals = true, SupportsOdbcTimestampLiterals = true ] , SQLGetFunctions = [ SQL_API_SQLBINDPARAMETER = false, SQL_CONVERT_FUNCTIONS = 2 ] , SQLGetInfo = [ SQL_SQL_CONFORMANCE = 8 ]  ] ","ProductVersion":"2.112.1161.0 (22.12)","ActivityId":"c6cbccaa-910e-42e5-98b3-d6b244facdcc","Process":"Microsoft.Mashup.Container.NetFX45","Pid":30036,"Tid":1,"Duration":"00:00:00.0000080"}
```

### CODE REVIEW

File: connector/Diagnostics.pqm [hunk 1]
```diff
@@ -0,0 +1,165 @@
+let
```
Comment (bucaojit) [inline connector/Diagnostics.pqm:1]: This file is from the connector samples and does not need review.

**Review by jchemburkar — APPROVED** [review]
This looks great! I had one (non-blocking) question, but otherwise LGTM

File: connector/MongoDBAtlasODBC.pq [hunk 2]
```diff
@@ -113,13 +121,50 @@ shared GetConnectionString = (uri as text, database as text) as record =>
         else
             ConnectionString;
 
+DataSourceMissingClientLibrary = "DataSource.MissingClientLibrary";
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply connector/MongoDBAtlasODBC.pq:125]: Sure, made the change.

File: connector/Diagnostics.pqm [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            try Serialize(value) otherwise "<serialization failed>"
+in
+    [ 
+        LogValue = Diagnostics.LogValue,
+        LogValue2 = Diagnostics.LogValue2,
+        LogFailure = Diagnostics.LogFailure,
+        WrapFunctionResult = Diagnostics.WrapFunctionResult,
+        WrapHandlers = Diagnostics.WrapHandlers,
+        ValueToText = Diagnostics.ValueToText
+    ]
```
Comment (nbagnard) [inline connector/Diagnostics.pqm:165]: [nit] Github is complaining about a missing new line at the end of the file

**Review by nbagnard — APPROVED** [review]
LGTM!

**Comment by bucaojit** [comment]
evergreen merge

**Pull request merged**